src/runtime/hexagon/hexagon_user_dma.cc

```diff
@@ -82,12 +95,13 @@ int HexagonUserDMA::Copy(void* dst, void* src, uint32_t length) {
     dmstart(dma_desc);
     first_dma_ = false;
   } else {
-    // `dmlink` descriptor to tail
-    dmlink(dma_descriptors_.back(), dma_desc);
+    // `dmlink` descriptor to tail descriptor
+    void* tail = GetDescriptorAddr(id_next_dma_desc_ - 1);
+    dmlink(tail, dma_desc);
   }
 
-  // set descriptor as new tail
-  dma_descriptors_.push_back(dma_desc);
+  // update the ID of the next DMA descriptor
+  id_next_dma_desc_++;
```

Comment:
Interaction with the ring buffer here feels a bit like it should be encapsulated into methods like "AllocDescriptor" "FreeDescriptor" or something like that which manages the internal member `id_next_dma_desc_` and others.

Reply:
Added new RingBuffer class along with testing.